pkg/statsd/handler_cloud.go

```diff
@@ -41,6 +50,92 @@ type CacheOptions struct {
+func constructCloudHandlerFactory(cloudProviderName string, logger logrus.FieldLogger, cacheOptions CacheOptions, limiter *rate.Limiter, version string) *CloudHandlerFactory {
+	return &CloudHandlerFactory{
+		CacheOptions:      &cacheOptions,
+		logger:            logger,
+		Limiter:           limiter,
+		CloudProvider:     nil,
+		cloudProviderName: cloudProviderName,
+		version:           version,
+	}
+}
+
+// ConstructCloudHandlerFactoryFromViper initialises a new cloud handler factory with defaults based on the name of the cloud
+// provider requested.
+func ConstructCloudHandlerFactoryFromViper(v *viper.Viper, logger logrus.FieldLogger, version string) (*CloudHandlerFactory, error) {
+	cloudProviderName := v.GetString(ParamCloudProvider)
+	if cloudProviderName == "" {
+		return constructCloudHandlerFactory(cloudProviderName, logger, CacheOptions{}, nil, version), nil
+	}
+
+	// Cloud provider defaults
+	cpDefaultCacheOpts, ok := DefaultCloudProviderCacheValues[cloudProviderName]
+	if !ok {
+		return nil, fmt.Errorf("could not find default cache values for cloud provider '%s'", cloudProviderName)
+	}
+	cpDefaultLimiterOpts, ok := DefaultCloudProviderLimiterValues[cloudProviderName]
+	if !ok {
+		return nil, fmt.Errorf("could not find default cache values for cloud provider '%s'", cloudProviderName)
+	}
+
+	// Set the defaults in Viper based on the cloud provider values before we manipulate things
+	v.SetDefault(ParamCacheRefreshPeriod, cpDefaultCacheOpts.CacheRefreshPeriod)
+	v.SetDefault(ParamCacheEvictAfterIdlePeriod, cpDefaultCacheOpts.CacheEvictAfterIdlePeriod)
+	v.SetDefault(ParamCacheTTL, cpDefaultCacheOpts.CacheTTL)
+	v.SetDefault(ParamCacheNegativeTTL, cpDefaultCacheOpts.CacheNegativeTTL)
+	v.SetDefault(ParamMaxCloudRequests, cpDefaultLimiterOpts.MaxCloudRequests)
+	v.SetDefault(ParamBurstCloudRequests, cpDefaultLimiterOpts.BurstCloudRequests)
+
+	// Set the used values based on the defaults merged with any overrides
+	cacheOptions := CacheOptions{
+		CacheRefreshPeriod:        v.GetDuration(ParamCacheRefreshPeriod),
+		CacheEvictAfterIdlePeriod: v.GetDuration(ParamCacheEvictAfterIdlePeriod),
+		CacheTTL:                  v.GetDuration(ParamCacheTTL),
+		CacheNegativeTTL:          v.GetDuration(ParamCacheNegativeTTL),
+	}
+	limiter := rate.NewLimiter(rate.Limit(v.GetInt(ParamMaxCloudRequests)), v.GetInt(ParamBurstCloudRequests))
+	return constructCloudHandlerFactory(cloudProviderName, logger, cacheOptions, limiter, version), nil
+}
+
+// InitCloudProvider initialises the cloud provider given some already parsed defaults.
+// This is mainly split out to enable testing the config parsing separately from the cloud providers.
+func (f *CloudHandlerFactory) InitCloudProvider(v *viper.Viper) error {
+	options := gostatsd.CloudProviderOptions{Viper: v, Logger: f.logger, Version: f.version}
+	switch f.cloudProviderName {
+	case aws.ProviderName:
+		// AWS has no special options required
+		break
```

Comment:
Don't need `break` here :)

Reply:
I think we do - unless I just remove that example clause. Otherwise if someone is using the AWS provider then we'll override them with k8s options. These options are harmless for now but it could introduce a bug down the road.

The two options I see here are:

1. keep it the way it is
2. remove the AWS clause completely since it's just a big comment, basically

Reply:
https://play.golang.org/p/pelegl1WgOt

Reply:
Ah. My bad for assuming, sorry Mikhail!